### PR TITLE
feat(MangaLife): add presence

### PR DESCRIPTION
**Pull request by ghost**
<details>
<summary> Proof that the created presence is working: </summary>
<br>

</details>
<details>
<summary> Some Notes for Presence.ts: </summary>
• Lines 4 to 13: The `let` allows for faster update of data instead of being updated only if you refreshed the webpage (especially when reading chapters, as chapters of manga are being dynamically loaded, I think). Also, it is a lot readable and maintainable like this, personally.<br>
• Lines 21 and 27: This is the same site being used by MangaLife to get their manga covers.
</details>

### CODE REVIEW

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Please look at other presences to see how they're structured, this is not to our repository standard.

**Comment by yrnmsk** [comment]
> Please look at other presences to see how they're structured, this is not to our repository standard.

do you mean the usage of let variables instead of `presenceData.<> = ...;` explicitly inside every if/else if code block?

if yes, then i noted that this was due to the slow update of data or a case where there is a need to refresh the whole page just to update the data. not really sure as to why and how it happens, but if it is still insisted then i will try to explore more so i can refactor.

if not, then can i know which presences should i maybe try take a look at? i did try to look at other presences when i was making this, but when i was in vs code, i couldnt really differentiate outdated and new presences so im not really sure of repo standards then.

**Comment by EncryptedDev** [comment]
> if not, then can i know which presences should i maybe try take a look at? i did try to look at other presences when i was making this, but when i was in vs code, i couldnt really differentiate outdated and new presences so im not really sure of repo standards then.

Use TruckersMP, LastFM and Netflix

**Comment by yrnmsk** [comment]
> Use TruckersMP, LastFM and Netflix

okay thankss, will do a quick refactor after checking those out

**Comment by yrnmsk** [comment]
> Use TruckersMP, LastFM and Netflix

**refactor: general recode**
refactored whole code according to repo standards using the three mentioned above as reference.
<br>
**feat: add settings** (privacy mode, cover, timestamps, buttons)
got inspired by the reference to finally implement and cross out one of my todo list (settings)
<details>
<summary> Proof that the new feature is working: </summary>
<br>

image

Privacy Mode (Off & On)
image
image

Show Cover (Off)
image

Show Timestamps (Off)
image

Show Buttons (Off)
image

</summary>

File: websites/M/MangaLife/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  "service": "MangaLife",
+  "altnames": [
+    "Manga4Life"
+  ],
+  "description": {
+    "en": "MangaLife: Read free manga online."
+  },
+  "url": "www.manga4life.com",
+  "version": "1.1.0",
+  "logo": "https://i.imgur.com/DvWL2co.png",
```
Comment (EncryptedDev) [inline websites/M/MangaLife/dist/metadata.json:16]: Must be 512x512

File: websites/M/MangaLife/presence.ts [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    details: "Browsing...",
+    state: "at Home Page",
+    startTimestamp: browsingStamp
+  };
+  const path = document.location.pathname,
+    pages: Record<string, PresenceData> = {
+      "/manga/": {
+        details: "Viewing a Manga...",
+        state: document.title.split(" | MangaLife")[0],
+        startTimestamp: browsingStamp
```
Comment (EncryptedDev) [inline websites/M/MangaLife/presence.ts:23]: This is set in the `presenceData` object on L11, you do not need to redefine this. Remove all matching instances too

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/M/MangaLife/dist/metadata.json (hunk 1 above)
Comment (yrnmsk) [reply websites/M/MangaLife/dist/metadata.json:16]: oops, i didn't notice, will fix asap.

File: websites/M/MangaLife/presence.ts (hunk 2 above)
Comment (yrnmsk) [reply websites/M/MangaLife/presence.ts:23]: also didn't notice, it was pretty late and i may have rushed a lot. will also fix asap.

**Comment by yrnmsk** [comment]
resolved requested changes with the new pushed changes.

asking again if the error in check schema validation is alright? i bumped it again from v.1.1.0 to v.1.1.1 because of the minor fixes and minor changes but the error bugs me on whether i should or shouldnt have done that

**Comment by Slowlife01** [comment]
> resolved requested changes with the new pushed changes.
> 
> asking again if the error in check schema validation is alright? i bumped it again from v.1.1.0 to v.1.1.1 because of the minor fixes and minor changes but the error bugs me on whether i should or shouldnt have done that. im still new to this semver so im a bit confused

Your presence hasn't been published, it doesn't make sense for you to bump version

**Comment by yrnmsk** [comment]
> Your presence hasn't been published, it doesn't make sense for you to bump version

so i should just put `1.0.0` as is? my bad, not really that familiar with semver yet. will change it rn

File: websites/M/MangaLife/presence.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  const [privacy, cover, timestamps, buttons] = await Promise.all([
+    presence.getSetting<boolean>("privacy"),
+    presence.getSetting<boolean>("cover"),
+    presence.getSetting<boolean>("timestamps"),
+    presence.getSetting<boolean>("buttons")
+  ]);
+  let presenceData: PresenceData = {
+    largeImageKey: "logo",
+    smallImageKey: !privacy ? "home" : "",
+    details: "Browsing...",
```
Comment (EncryptedDev) [inline websites/M/MangaLife/presence.ts:14]: Browsing homepage instead of state and details here

File: websites/M/MangaLife/presence.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      presenceData.largeImageKey =
+        !privacy && cover
+          ? `https://cover.nep.li/cover/${path.split("/manga/")[1]}.jpg`
+          : "logo";
+      presenceData.smallImageKey = !privacy && cover ? "logo-png" : "search";
+      presenceData.buttons = [
+        {
+          label:
+            presenceData.state.length >= 30 ? "View Manga" : presenceData.state,
+          url: `${document.location}`
```
Comment (EncryptedDev) [inline websites/M/MangaLife/presence.ts:86]:
```suggestion
          url: document.location
```

File: websites/M/MangaLife/presence.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            presenceData.details.length >= 30
+              ? "View Manga"
+              : presenceData.details,
+          url: `${document.location.origin}/manga/${
+            path.split("/read-online/")[1].split("-chapter-")[0]
+          }`
+        },
+        {
+          label: `${presenceData.state.replace("|", " ")}`,
+          url: `${document.location}`
```
Comment (EncryptedDev) [inline websites/M/MangaLife/presence.ts:117]: Ditto

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

File: websites/M/MangaLife/presence.ts (hunk 3 above)
Comment (yrnmsk) [reply websites/M/MangaLife/presence.ts:14]: but i kinda reused the `details: Browsing...` for several other pages, so changing it from `Browsing...` to `Browsing Home Page...` can be a conflict

File: websites/M/MangaLife/presence.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    presence.getSetting<boolean>("privacy"),
+    presence.getSetting<boolean>("cover"),
+    presence.getSetting<boolean>("timestamps"),
+    presence.getSetting<boolean>("buttons")
+  ]);
+  let presenceData: PresenceData = {
+    largeImageKey: "logo",
+    smallImageKey: !privacy ? "home" : "",
+    details: "Browsing...",
+    state: "at Home Page",
```
Comment (EncryptedDev) [inline websites/M/MangaLife/presence.ts:15]:
```suggestion
    state: "Home Page",
```

File: websites/M/MangaLife/presence.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    startTimestamp: browsingStamp
+  };
+  const path = document.location.pathname,
```
Comment (EncryptedDev) [inline websites/M/MangaLife/presence.ts:18]: Merge this with the constants above the let

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

**Comment by yrnmsk** [comment]
resolved requested changes with the new pushed changes, took a little while as i was busy with other stuff.

**Comment by ImRodry** [comment]
This doesn't need a rebase but it should be rebased anyway

**Review by Bas950 — APPROVED** [review]

**Review by EncryptedDev — APPROVED** [review]

**Pull request merged**